components/scheduler2/generator_executors.go

```diff
@@ -0,0 +1,163 @@
+package scheduler2
+
+import (
+	"fmt"
+	"time"
+
+	"go.temporal.io/api/serviceerror"
+	servercommon "go.temporal.io/server/common"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/service/history/hsm"
+	"go.uber.org/fx"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type (
+	GeneratorTaskExecutorOptions struct {
+		fx.In
+
+		Config         *Config
+		MetricsHandler metrics.Handler
+		Logger         log.Logger
+		SpecProcessor  *SpecProcessor
+	}
+
+	generatorTaskExecutor struct {
+		GeneratorTaskExecutorOptions
+	}
+)
+
+func RegisterGeneratorExecutors(registry *hsm.Registry, options GeneratorTaskExecutorOptions) error {
+	e := generatorTaskExecutor{options}
+	if err := hsm.RegisterTimerExecutor(registry, e.executeBufferTask); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (e generatorTaskExecutor) executeBufferTask(env hsm.Environment, node *hsm.Node, task BufferTask) error {
+	schedulerNode := node.Parent
+	scheduler, err := e.loadScheduler(schedulerNode)
+	if err != nil {
+		return err
+	}
+
+	generator, err := e.loadGenerator(node)
+	if err != nil {
+		return err
+	}
+
+	// If we have no last processed time, this is a new schedule.
+	if generator.LastProcessedTime == nil {
+		generator.LastProcessedTime = timestamppb.Now()
+		// TODO - update schedule info with create time
+
+		e.logSchedule("Starting schedule", scheduler)
+	}
+
+	// Process time range between last high water mark and system time.
+	t1 := generator.LastProcessedTime.AsTime()
+	t2 := time.Now().UTC()
```

Comment:
Use `env.Now()` instead, this will potentially allow you to inject a fake time source and fast forward for testing purposes.